File: goldencheck/semantic/suppression.py
"""Suppression engine — downgrades irrelevant findings based on column semantic type."""
from __future__ import annotations
from dataclasses import replace
from goldencheck.models.finding import Finding, Severity
from goldencheck.semantic.classifier import ColumnClassification, TypeDef

def apply_suppression(
    findings: list[Finding],
    column_types: dict[str, ColumnClassification],
    type_defs: dict[str, TypeDef],
) -> list[Finding]:
    """Downgrade findings where check is in the column type's suppress list."""
    result = []
    for f in findings:
        # Only suppress WARNING/ERROR
        if f.severity not in (Severity.WARNING, Severity.ERROR):
            result.append(f)
            continue
        # Never suppress LLM findings
        if f.source == "llm":
            result.append(f)
            continue
        # Never suppress high initial confidence (>= 0.9)
        if f.confidence >= 0.9:
            result.append(f)
            continue

        # Check if this finding should be suppressed
        classification = column_types.get(f.column)
        if classification and classification.type_name:
            type_def = type_defs.get(classification.type_name)
            if type_def and f.check in type_def.suppress:
                # For pattern_consistency findings on geo/identifier columns,
                # skip suppression if patterns are code-like (mostly digits)
                # and differ significantly in length (e.g. 5-digit vs 9-digit
                # zip codes). This preserves real format inconsistencies while
                # still suppressing natural variation in names/cities.
                if (
                    f.check == "pattern_consistency"
                    and classification.type_name in ("geo", "identifier")
                ):
                    dom = f.metadata.get("dominant_pattern")
                    minor = f.metadata.get("minority_pattern")
                    if dom and minor and abs(len(dom) - len(minor)) > 1:
                        # Only skip suppression when BOTH patterns are
                        # code-like (mostly digits). Mixed patterns (e.g.
                        # DD vs LLLLLLL in age columns) should stay suppressed.
                        dom_digit = sum(1 for c in dom if c == "D") / max(len(dom), 1)
                        minor_digit = sum(1 for c in minor if c == "D") / max(len(minor), 1)
                        if dom_digit > 0.5 and minor_digit > 0.5:
                            result.append(f)
                            continue
                result.append(replace(
                    f,
                    severity=Severity.INFO,
                    message=f"{f.message} (suppressed: {classification.type_name} column)",
                ))
                continue

        result.append(f)
    return result

File: packages/goldencheck-js/src/core/semantic/suppression.ts
/**
 * Suppression engine — downgrades findings based on semantic type.
 * Port of goldencheck/semantic/suppression.py.
 */

import type { Finding, TypeDef, ColumnClassification } from "../types.js";
import { Severity, replaceFinding } from "../types.js";

/**
 * Apply type-based suppression to findings.
 * Matching checks are downgraded to INFO (never suppresses LLM or high-confidence).
 */
export function applySuppression(
  findings: readonly Finding[],
  columnTypes: Readonly<Record<string, ColumnClassification>>,
  typeDefs: Readonly<Record<string, TypeDef>>,
): Finding[] {
  return findings.map((f) => {
    // Only suppress WARNING/ERROR
    if (f.severity === Severity.INFO) return f;

    // Never suppress LLM findings
    if (f.source === "llm") return f;

    // Never suppress high-confidence findings
    if (f.confidence >= 0.9) return f;

    const classification = columnTypes[f.column];
    if (!classification || !classification.typeName) return f;

    const typeDef = typeDefs[classification.typeName];
    if (!typeDef) return f;

    if (!typeDef.suppress.includes(f.check)) return f;

    // Special handling for pattern_consistency on geo/identifier types
    if (f.check === "pattern_consistency" && (classification.typeName === "geo" || classification.typeName === "identifier")) {
      const dominantPattern = f.metadata["dominant_pattern"] as string | undefined;
      const minorityPattern = f.metadata["minority_pattern"] as string | undefined;

      if (dominantPattern && minorityPattern) {
        const lengthDiff = Math.abs(dominantPattern.length - minorityPattern.length);
        const domDigitRatio = countChar(dominantPattern, "D") / Math.max(dominantPattern.length, 1);
        const minDigitRatio = countChar(minorityPattern, "D") / Math.max(minorityPattern.length, 1);

        // Preserve real format issues (e.g., 5-digit vs 9-digit zip)
        if (lengthDiff > 1 && domDigitRatio > 0.5 && minDigitRatio > 0.5) {
          return f; // Skip suppression
        }
      }
    }

    // Suppress: downgrade to INFO
    return replaceFinding(f, {
      severity: Severity.INFO,
      message: `${f.message} (suppressed: ${classification.typeName} column)`,
    });
  });
}

function countChar(s: string, ch: string): number {
  let count = 0;
  for (const c of s) {
    if (c === ch) count++;
  }
  return count;
}
